src/oscar/apps/offer/abstract_models.py

```diff
@@ -973,31 +973,30 @@ def product_queryset(self):
             return Product.objects.exclude(
                 id__in=self.excluded_products.values("id")
             )
-
+        # start with filter clause that always applies
+        _filter = Q(includes=self)
+        # extend filter if included classes exist:
+        if self.classes.exists():
+            _filter |= Q(product_class__classes=self)
+            _filter |= Q(parent__product_class__classes=self)
+        # extend filter if included_products have children
+        if Product.objects.filter(parent__includes=self).exists():
+            _filter |= Q(parent__includes=self)
+        # extend filter if included_categories exist
         if self.included_categories.exists():
             expanded_range_categories = ExpandDownwardsCategoryQueryset(
                 self.included_categories.values("id")
             )
-            selected_products = Product.objects.filter(
-                Q(categories__in=expanded_range_categories)
-                | Q(product_class__classes=self)
-                | Q(includes=self)
-                | Q(parent__categories__in=expanded_range_categories)
-                | Q(parent__product_class__classes=self)
-                | Q(parent__includes=self),
-                ~Q(excludes=self),
-                ~Q(parent__excludes=self)
-            )
-        else:
-            selected_products = Product.objects.filter(
-                Q(product_class__classes=self)
-                | Q(includes=self)
-                | Q(parent__product_class__classes=self)
-                | Q(parent__includes=self),
-                ~Q(excludes=self),
-                ~Q(parent__excludes=self)
-            )
-        return selected_products.distinct()
+            _filter |= Q(categories__in=expanded_range_categories)
+            # extend filter for parent categories
+            if Product.objects.filter(
+                    parent__categories__in=expanded_range_categories).exists():
```

Comment:
I think you can save 1 extra query by doing the following:. Save the value returned by Product.objects.filter(parent__includes=self).exists() in a variable and if there are child products, it makes sence to add this query, if there are no children you don;t need to do it.

In the same way you can use this variable to descide if you need to add ``parent__excludes`` to the query.

Do you think this is correct?